File: ma_sh/Cpp/src/filter.cpp
#include "filter.h"

using namespace torch::indexing;

const torch::Tensor toMaxValues(const int &unique_idx_num,
                                const torch::Tensor &data,
                                const torch::Tensor &data_idxs) {
  std::vector<torch::Tensor> max_values_vec;
  max_values_vec.reserve(unique_idx_num);

  for (int i = 0; i < unique_idx_num; ++i) {
    const torch::Tensor current_max_value =
        torch::max(data.index({data_idxs == i}));

    max_values_vec.emplace_back(current_max_value);
  }

  const torch::Tensor max_values = torch::hstack(max_values_vec);

  return max_values;
}

File: ma_sh/Method/Mash/filter.py
import torch


def toMaxValues(data: torch.Tensor, data_idxs: torch.Tensor) -> torch.Tensor:
    max_values_list = []

    unique_idxs = torch.unique(data_idxs)

    for i in range(unique_idxs.shape[0]):
        current_idx = unique_idxs[i]

        current_max_value = torch.max(data[data_idxs == current_idx])

        max_values_list.append(current_max_value)

    max_values = torch.hstack(max_values_list)

    return max_values
